Embassy/Logging/DefaultLogger.swift

```diff
@@ -0,0 +1,112 @@
+    let name: String
+    let level: LogLevel
+    private(set) var handlers: [LogHandler] = []
+
+    public init(name: String, level: LogLevel = .info) {
+        self.name = name
+        self.level = level
+    }
+
+    public init(fileName: String = #file, level: LogLevel = .info) {
+        self.name = DefaultLogger.moduleNameForFileName(fileName)
+        self.level = level
+    }
+
+    /// Add handler to self logger
+    ///  - Parameter handler: the handler to add
+    public func addHandler(_ handler: LogHandler) {
+        handlers.append(handler)
+    }
+
+    public func debug(
+        _ message: @autoclosure (Void) -> String,
+        caller: String = #function,
+        file: String = #file,
+        line: Int = #line
+    ) {
+        log(.debug, message: message, caller: caller, file: file, line: line)
+    }
+
+    public func info(
+        _ message: @autoclosure (Void) -> String,
+        caller: String = #function,
+        file: String = #file,
+        line: Int = #line
+    ) {
+        log(.info, message: message, caller: caller, file: file, line: line)
+    }
+
+    public func warning(
+        _ message: @autoclosure (Void) -> String,
+        caller: String = #function,
+        file: String = #file,
+        line: Int = #line
+    ) {
+        log(.warning, message: message, caller: caller, file: file, line: line)
+    }
+
+    public func error(
+        _ message: @autoclosure (Void) -> String,
+        caller: String = #function,
+        file: String = #file,
+        line: Int = #line
+    ) {
+        log(.error, message: message, caller: caller, file: file, line: line)
+    }
+
+    public func critical(
+        _ message: @autoclosure (Void) -> String,
+        caller: String = #function,
+        file: String = #file,
+        line: Int = #line
+    ) {
+        log(.critical, message: message, caller: caller, file: file, line: line)
+    }
+
+    func log(
+        _ level: LogLevel,
+        message: @autoclosure (Void) -> String,
+        caller: String = #function,
+        file: String = #file,
+        line: Int = #line
+    ) {
+        let record = LogRecord(
+            loggerName: name,
+            level: level,
+            message: message(),
+            file: file,
+            function: caller,
+            line: line,
+            time: Date()
+        )
+        log(record)
+    }
+
+    public func log(_ record: LogRecord) {
+        guard record.level.rawValue >= level.rawValue else {
+            return
+        }
+        for handler in handlers {
+            handler.emit(record)
+        }
+    }
+
+    /// Strip file name and return only the name part, e.g. /path/to/MySwiftModule.swift will be
+    /// MySwiftModule
+    ///  - Parameter fileName: file name to be stripped
+    ///  - Returns: stripped file name
+    static func moduleNameForFileName(_ fileName: String) -> String {
+        return ((fileName as NSString).lastPathComponent as NSString).deletingPathExtension
```

Comment:
hmmm, maybe we should get rid of NSString here, as it will stop this project from compiling under non-iOS environment, say like linux